### PR TITLE
Simplify input helper

### PR DESCRIPTION
**Pull request by drogus**
The motivation behind this PR is to simplify the `input` helper, so that `TextField` and `Checkbox` subclasses can be rendered directly and work just like `input`. Having additional code in the `input` helper may make it confusing at the moment. For example with a component `CustomTextField = Ember.TextField.extend()` the following code will not work as expected (what's worse it will throw a non descriptive error):

``` handlebars
{{custom-text-field value=something action="doSomething" on="key-up"}}
```

@rwjblue mentioned that I shouldn't override the `create` method on `TextField`, but unfortunately I can't see any way to do that. If that's a no go, I'd like to propose deprecating `on` shortcut and either use `on-event` or come up with something new.

### CODE REVIEW

**Comment by mmun** [comment]
I like the direction of moving the logic out of the helpers and into create. Helpers should be a fairly thing layer. Ideally this helper wouldn't depend on calling the viewHelper either.

**Comment by drogus** [comment]
> Ideally this helper wouldn't depend on calling the viewHelper either.

Could you say more? I'm curious how it could be achieved.

**Comment by rwjblue** [comment]
@drogus - The needed functionality of the `{{view}}` helper has been extracted to a couple functions. At this point it is equivalent to following (for the `{{input}}` case) now:

```javascript
import mergeViewBindings from "ember-htmlbars/system/merge-view-bindings";
import appendTemplatedView from "ember-htmlbars/system/append-templated-view";

// then instead of invoking the view helper do the following:

  mergeViewBindings(this, props, hash);
  appendTemplatedView(this, options.morph, viewClassOrInstance, props);
```

File: packages/ember-views/lib/views/text_field.js [hunk 1]
```diff
@@ -111,3 +111,20 @@ export default Component.extend(TextSupport, {
...
+    attributes = attributes || {};
+
+    Ember.assert('You can either pass `on` or `onEvent` to TextField, not both', !(attributes.on && attributes.onEvent));
+
+    if (!attributes.onEvent) {
+      attributes.onEvent = attributes.on || 'enter';
+      delete attributes.on;
+    }
+
+    return this._super(attributes);
```
Comment (stefanpenner) [inline packages/ember-views/lib/views/text_field.js]: `this._super.apply(this, arguments)`

**Comment by drogus** [comment]
@rwjblue thanks, I've changed the implementation to use that instead of a view helper

**Comment by tchak** [comment]
@drogus would be nice if in the process we could use `Component` instead of `View` in `Checkbox`. Maybe it is a different issue.

**Comment by lolmaus** [comment]
> For example with a component CustomTextField = Ember.TextField.extend() the following code will not work as expected (what's worse it will throw a non descriptive error):
> 
>    {{custom-text-field value=something action="doSomething" on="key-up"}}

Huh?

Here's what i'm doing:

```coffee
`import Ember from 'ember'`

SelfFocusingTextFieldComponent = Ember.TextField.extend

  # Arguments
  valid: true

  # Options
  classNames: ['self-focusing-text-field']
  classNameBindings: ['valid:-valid:-invalid']
  instrumentDisplay: '{{self-focusing-text-field}}'
  
  # Built-in methods
  didInsertElement: ->
    @$().focus()

`export default SelfFocusingTextFieldComponent`
```

And here's how i'm using it:

```hbs
<h2 class="selectionSlim-title">

  {{#if isEditingTitle}}
    {{self-focusing-text-field
      value=selection.name
      focus-out="finishEditingTitle"
      insert-newline="finishEditingTitle"}}
  {{else}}
    ...
  {{/if}}
</h2>
```

It works without any errors.

**Comment by lolmaus** [comment]
Can you write a unit test that fails on your use case with the current TextField implementation?

**Comment by rwjblue** [comment]
> we could use Component instead of View in Checkbox. Maybe it is a different issue.

@tchak - Yes, that is a separate open issue.

**Comment by drogus** [comment]
> It works without any errors.

It will fail if you use `on="some-event"`

> we could use Component instead of View in Checkbox. Maybe it is a different issue.

I was thinking about that too, I'll submit a separate PR for it.

**Comment by drogus** [comment]
> Can you write a unit test that fails on your use case with the current TextField implementation?

This test will fail if you revert my changes: https://github.com/emberjs/ember.js/pull/10283/files#diff-e8db379716498a52ea78f3e3cd710589R208

**Comment by rwjblue** [comment]
> I was thinking about that too, I'll submit a separate PR for it.

There have been more than one PR to fix this, it is not a SemVer compatible change.

**Comment by tchak** [comment]
@rwjblue so componentizing `Checkbox` and `Select` is out of reach before Ember 2.0, right ?

**Comment by jonkoops** [comment]
@tchak `Checkbox` and `Select` could be introduced as components we could also deprecate the view versions and remove them in a later point in time. It's probably a good idea to make them 'data down actions up' since as far as I know the view versions only support observation of the bound property.

**Comment by stefanpenner** [comment]
as of 1.13.3 we can actually just interact with native DOM as is.

result in:
```hbs
<input type="range" 
       value={{currentValue}}
       min=1
       max=10
       oninput={{action "changed" value="target.value"}} >
       

    <input type="checkbox" 
       value={{currentValue}}
       min=1
       max=10
       onchange={{action "changed" value="target.checked"}} >
       
     <input type="text" 
       value={{currentValue}}
       min=1
       max=10
       oninput={{action "changed" value="target.value"}} >
       
      <form action="" onchange={{action "changed" value="target.value"}} >
         <input type="radio"
                name="sex"
                selected={{is-equal currentValue "friend"}}
                value="friend">friend
         <input type="radio"
                name="sex"
                value="foe"
                selected={{is-equal currentValue "friend"}}>
                Foe
      </form>
```

in light of this, do we still want this PR?

**Comment by stefanpenner** [comment]
i suspect the subclassing story will be to be explored...

**Comment by rwjblue** [comment]
The `on` craziness was replaced by [an AST transform](https://github.com/emberjs/ember.js/blob/master/packages/ember-template-compiler/lib/plugins/transform-input-on-to-onEvent.js) that moves `on` to `onEvent`.  [The keyword](https://github.com/emberjs/ember.js/blob/b3b6e474b8074c77ba82e7d2b3bf1204770af1d8/packages/ember-htmlbars/lib/keywords/input.js#L155-L171) is only used to allow the non-dashed `{{input}}` to work.

@drogus - Thank you for working so hard on this!

**Pull request closed without merging**